### PR TITLE
add dump bytecode debug mode

### PR DESCRIPTION
**Pull request by pinzart90**
Moved DebugModes code in DynamoUtilities (so that they can be referenced in low level assemblies)
Added a new Debug mode to trigger creation of bytecode logs on each Graph execution
Logs will be in the Current Working directory/ByteCodeLogs/somelog.log

### CODE REVIEW

File: src/Engine/ProtoScript/Runners/LiveRunner.cs [hunk 1]
```diff
@@ -1462,26 +1512,29 @@ private void SynchronizeInternal(GraphSyncData syncData)
         {
             runnerCore.Options.IsDeltaCompile = true;
 
-            if (syncData == null)
+            using (DebugModes.IsEnabled("DumpByteCode") ? new DebugByteCodeMode(runnerCore) : null)
```
Comment (mjkkirschner) [inline src/Engine/ProtoScript/Runners/LiveRunner.cs:1515]: does `using null` do anything?

Comment (pinzart90) [reply src/Engine/ProtoScript/Runners/LiveRunner.cs:1515]: MSDN says that the compiler translates the using statement into a try catch finally with a null check in the finally block
`finally
    {
        reader?.Dispose();
    }`
    
    https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-statement

Comment (pinzart90) [reply src/Engine/ProtoScript/Runners/LiveRunner.cs:1515]: Also `If a null resource is acquired, then no call to Dispose is made, and no exception is thrown`
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#the-using-statement

File: src/Engine/ProtoScript/Runners/LiveRunner.cs [hunk 2]
```diff
@@ -1028,6 +1031,52 @@ public interface ILiveRunner
...
+            private ExecutionMode oldExecutionMode;
+
+            private Core core;
+
+            internal DebugByteCodeMode(Core core)
+            {
+                // Setup the log file location
+                var logFolder = Path.Combine(Directory.GetCurrentDirectory(), "ByteCodeLogs");
+                Directory.CreateDirectory(logFolder);
+                var newWriter = File.CreateText(Path.Combine(logFolder, DateTime.Now.ToString("yyyy-MM-ddTHH-mm-ss-fff") + ".log"));
```
Comment (mjkkirschner) [inline src/Engine/ProtoScript/Runners/LiveRunner.cs:1048]: should this be wrapped in a try catch?

Comment (mjkkirschner) [reply src/Engine/ProtoScript/Runners/LiveRunner.cs:1048]: so to enable or disable logging you need to turn this flag on(or off) and then reboot Dynamo?

**Review by mjkkirschner — APPROVED** [review]
It looks fine to me @pinzart90 - two questions though.

File: src/Engine/ProtoScript/Runners/LiveRunner.cs (hunk 2 above)
Comment (pinzart90) [reply src/Engine/ProtoScript/Runners/LiveRunner.cs:1048]: no....you can start dynamo and enable the debug mode to start generating logs.
no need for a reboot.

Not sure a try catch is worth it...this is going to be used by devs to debug

Comment (mjkkirschner) [reply src/Engine/ProtoScript/Runners/LiveRunner.cs:1048]: ok so `SynchronizeInternal` gets called a lot - I guess this means we might have log files per each run - that seems good. 👍

**Pull request merged**